src/components/Trade/hooks/useSwapper/useSwapper.ts

```diff
@@ -50,157 +54,238 @@ export enum TRADE_ERRORS {
+    wallet
+  }: {
+    wallet: HDWallet
+  }): Promise<ExecQuoteOutput | undefined> => {
+    let result
+    try {
+      const swapper = swapperManager.getSwapper(bestSwapperType)
+      result = await swapper.executeQuote({ quote, wallet })
+    } catch (err) {
+      setError('useSwapper', { message: TRADE_ERRORS.NO_LIQUIDITY })
+      console.error(`TradeProvider - executeQuote error: ${err}`) // eslint-disable-line no-console
+    }
+    return result
+  }
+
   const getQuoteFromSwapper = async <C extends ChainTypes, S extends SwapperType>({
     amount,
     sellAsset,
     buyAsset,
+    action,
     onFinish
-  }: GetQuote<C, S>) => {
+  }: GetQuoteFromSwapper<C, S>) => {
     if (debounceObj?.cancel) debounceObj.cancel()
     clearErrors()
     const quoteDebounce = debounce(async () => {
-      try {
-        const swapper = swapperManager.getSwapper(bestSwapperType)
-        let convertedAmount = amount
-        const isFiat = Object.keys(amount)[0].includes('fiat')
-        if (isFiat) {
-          const rate = await swapper.getUsdRate({
+      if (isComponentMounted.current) {
+        try {
+          const swapper = swapperManager.getSwapper(bestSwapperType)
+          let convertedAmount =
+            action === TradeActions.BUY ? { buyAmount: amount } : { sellAmount: amount }
+
+          const isFiat = action === TradeActions.FIAT
+          if (isFiat) {
+            const rate = await swapper.getUsdRate({
+              symbol: sellAsset.symbol,
+              tokenId: sellAsset.tokenId
+            })
+            convertedAmount = {
+              sellAmount: toBaseUnit(
+                bn(amount).div(rate).toString(),
+                sellAsset.precision
+              ).toString()
+            }
+          }
+          const quoteInput = {
+            sellAsset: sellAsset,
+            buyAsset: buyAsset,
+            ...convertedAmount
+          }
+
+          const { trade } = getValues()
+          let minMax = trade
+          if (
+            quote?.sellAsset?.symbol !== sellAsset.symbol &&
+            quote?.buyAsset?.symbol !== buyAsset.symbol
+          ) {
+            minMax = await swapper.getMinMax(quoteInput)
+            const minMaxTrade = { ...minMax, ...trade }
+            minMax && setValue('trade', minMaxTrade)
+          }
+          const newQuote = await swapper.getQuote({ ...quoteInput, ...minMax })
+          if (!newQuote?.success) throw newQuote
+
+          const sellAssetUsdRate = await swapper.getUsdRate({
             symbol: sellAsset.symbol,
             tokenId: sellAsset.tokenId
           })
-          const fiatAmount = Object.values(amount)[0]
-          convertedAmount = {
-            sellAmount: toBaseUnit(
-              bn(fiatAmount).div(rate).toString(),
-              sellAsset.precision
-            ).toString()
+          let buyAssetUsdRate
+          if (newQuote?.rate) {
+            buyAssetUsdRate = bn(sellAssetUsdRate).dividedBy(newQuote?.rate).toString()
```

Comment:
Make sure you are avoiding divide by zero errors and NaN values by guarding on the variables you expect to be valid numeric values.

```suggestion
            buyAssetUsdRate = bnOrZero(sellAssetUsdRate).div(newQuote?.rate || 1).toString()
```